packages/editor/src/components/post-taxonomies/hierarchical-term-selector.js

```diff
@@ -43,329 +34,325 @@ const DEFAULT_QUERY = {
+		// to have false values filtered out.
+		return false;
+	};
+	return matchTermsForFilter;
+}
+
+/**
+ * Hierarchical term selector.
+ *
+ * @param {Object} props      Component props.
+ * @param {string} props.slug Taxonomy slug.
+ * @return {WPElement}        Hierarchical term selector component.
+ */
+function HierarchicalTermSelector( { slug } ) {
+	const [ adding, setAdding ] = useState( false );
+	const [ formName, setFormName ] = useState( '' );
+	/**
+	 * @type {[number|'', Function]}
+	 */
+	const [ formParent, setFormParent ] = useState( '' );
+	const [ showForm, setShowForm ] = useState( false );
+	const [ filterValue, setFilterValue ] = useState( '' );
+	const [ filteredTermsTree, setFilteredTermsTree ] = useState( [] );
+	const debouncedSpeak = useDebounce( speak, 500 );
+
+	const {
+		hasCreateAction,
+		hasAssignAction,
+		terms,
+		loading,
+		availableTerms,
+		taxonomy,
+	} = useSelect(
+		( select ) => {
+			const { getCurrentPost, getEditedPostAttribute } = select(
+				editorStore
 			);
-		} );
+			const { getTaxonomy, getEntityRecords, isResolving } = select(
+				coreStore
+			);
+			const _taxonomy = getTaxonomy( slug );
+
+			return {
+				hasCreateAction: _taxonomy
+					? get(
+							getCurrentPost(),
+							[
+								'_links',
+								'wp:action-create-' + _taxonomy.rest_base,
+							],
+							false
+					  )
+					: false,
+				hasAssignAction: _taxonomy
+					? get(
+							getCurrentPost(),
+							[
+								'_links',
+								'wp:action-assign-' + _taxonomy.rest_base,
+							],
+							false
+					  )
+					: false,
+				terms: _taxonomy
+					? getEditedPostAttribute( _taxonomy.rest_base )
+					: EMPTY_ARRAY,
+				loading: isResolving( 'getEntityRecords', [
+					'taxonomy',
+					slug,
+					DEFAULT_QUERY,
+				] ),
+				availableTerms:
+					getEntityRecords( 'taxonomy', slug, DEFAULT_QUERY ) ||
+					EMPTY_ARRAY,
+				taxonomy: _taxonomy,
+			};
+		},
+		[ slug ]
+	);
+
+	const { editPost } = useDispatch( editorStore );
+	const { saveEntityRecord } = useDispatch( coreStore );
+
+	const availableTermsTree = useMemo(
+		() => sortBySelected( buildTermsTree( availableTerms ), terms ),
+		// Remove `terms` from the dependency list to avoid reordering every time
+		// checking or unchecking a term.
+		[ availableTerms ]
+	);
+
+	if ( ! hasAssignAction ) {
+		return null;
 	}
 
-	onAddTerm( event ) {
+	/**
+	 * Append new term.
+	 *
+	 * @param {Object} term Term object.
+	 * @return {Promise} A promise that resolves to save term object.
+	 */
+	const addTerm = ( term ) => {
+		return saveEntityRecord( 'taxonomy', slug, term );
+	};
+
+	/**
+	 * Update terms for post.
+	 *
+	 * @param {number[]} termIds Term ids.
+	 */
+	const onUpdateTerms = ( termIds ) => {
+		editPost( { [ taxonomy.rest_base ]: termIds } );
+	};
+
+	/**
+	 * Handler for checking term.
+	 *
+	 * @param {number} termId
+	 */
+	const onChange = ( termId ) => {
+		const hasTerm = terms.indexOf( termId ) !== -1;
```

Comment:
```suggestion
		const hasTerm = terms.includes( termId );
```

I think we can use `Array.includes()` here.